pkgs/os-specific/darwin/signing-utils/utils.sh

```diff
@@ -0,0 +1,43 @@
+# Work around for some odd behaviour where we can't codesign a file
+# in-place if it has been called before. This happens for example if
+# you try to fix-up a binary using strip/install_name_tool, after it
+# had been used previous.  The solution is to copy the binary (with
+# the corrupted signature from strip/install_name_tool) to some
+# location, sign it there and move it back into place.
+#
+# This does not appear to happen with the codesign tool that ships
+# with recent macOS BigSur installs on M1 arm64 machines.  However it
+# had also been happening with the tools that shipped with the DTKs.
+sign() {
+    local tmpdir
+    tmpdir=$(mktemp -d)
+
+    # $1 is the file
+
+    cp "$1" "$tmpdir"
+    CODESIGN_ALLOCATE=@codesignAllocate@ \
+        @sigtool@/bin/codesign -f -s - "$tmpdir/$(basename "$1")"
+    mv "$tmpdir/$(basename "$1")" "$1"
+    rmdir "$tmpdir"
+}
+
+checkRequiresSignature() {
+    local file=$1
+    local rc=0
+
+    @sigtool@/bin/sigtool --file "$file" check-requires-signature || rc=$?
```

Comment:
This bails on fat binaries, leaving them with a broken signature after stripping:

```
$ file /tmp/libgcc_s.2.dylib
/tmp/libgcc_s.2.dylib: Mach-O universal binary with 1 architecture: [arm64:Mach-O 64-bit dynamically linked shared library arm64]
/tmp/libgcc_s.2.dylib (for architecture arm64):	Mach-O 64-bit dynamically linked shared library arm64
```

Reply:
To add to reasons that we might want to stick to `codesign`:

* Build process might assume its presence and behavior (emacs)
* Build process might want to add entitlements (qemu)

Reply:
It would be nice to use upstream `codesign` instead of using [my reimplementation](https://github.com/thefloweringash/sigtool), but packaging the open source release is remarkably difficult. It was briefly in nixpkgs but reverted and the commit message isn't clear on why, some context at #38865 and #37838.

Reply:
> we _could_ work around the problem by extracting the thin library from the universal library

Now that I check your PR, I see that's exactly what you did. Sorry for the overexplain.

I've added a branch supporting signatures on universal binaries and I'm building your gcc PR to test.

Reply:
A related concern here I've discovered locally - some binaries, for example QEMU will require specific entitlements, in their case `com.apple.security.hypervisor`. I haven't looked at how entitlements are added to a binary, but if we don't want to use `codesign` it would be something `sigtool` needs to support. Ideally we would copy the entitlements from the broken signature.